src/cdomains/intDomain.ml

```diff
@@ -349,6 +349,446 @@ struct
     | None -> None
 end
 
+module IntPair =
+struct
+  type t = (int64 * int64)
+
+  let compare (x0, y0) (x1, y1) =
+    match Int64.compare x0 x1 with
+    | 0 -> Int64.compare y0 y1
+    | c -> c
+
+  let leq (x0, y0) (x1, y1) : bool =
+    Int64.compare x0 y0 >= 0 && Int64.compare x1 y1 <= 0
+end
+
+(* Most of functions implemented here basically work the same way as their respective ones from
+   Interval32, so they will be commented only if they differ significantly from the Interval32 ones*)
+module IntervalSet : S with type t = Set.Make(IntPair).t option =
+struct
+  include Printable.Std
+
+  open Int64
+
+  module SS = Set.Make(IntPair)
+  type t = SS.t option
+
+  let min_int, max_int = Size.range Cil.IInt (* TODO this currently depends on the machine Cil was compiled on... *)
+
+  let to_yojson _ = failwith "TODO to_yojson"
+
+  let top () = Some (SS.add (min_int, max_int) SS.empty)
+  let bot () = None
+  let is_top x = x = top()
+  let is_bot = function None -> true | _ -> false
+  let zero = Some (SS.add (0L, 0L) SS.empty)
+  let one = Some (SS.add (1L, 1L) SS.empty)
+  let top_bool = Some (SS.add (0L, 1L) SS.empty)
+
+  let hash (x:t) = Hashtbl.hash x
+  let short _ = function
+  | None -> "bottom"
+  | Some s ->
+    let s_string = SS.elements s |> List.sort IntPair.compare |>
+        List.fold_left (fun str (x1, x2) -> str ^ "[" ^ to_string x1 ^ "," ^ to_string x2 ^ "], ") "" in
+    String.sub s_string 0 (String.length s_string - 2)
+
+  let isSimple _ = true
+  let name () = "set of 32bit intervals"
+  let pretty_f sh () x = text (sh 80 x)
+  let pretty = pretty_f short
+  let toXML_f sh x = Xml.Element ("Leaf", [("text", sh 80 x)], [])
+  let toXML = toXML_f short
+  let printXml f x = BatPrintf.fprintf f "<value>\n<data>\n%s\n</data>\n</value>\n" (short 800 x)
+  let pretty_diff () (x,y) = Pretty.dprintf "%a instead of %a" pretty x pretty y
+
+  (* meet_lists takes sorted lists of intervals. The algorithm works as follows:
+     - if selected intervals do not intersect, truncate the one with the least left border and proceed to the next one
+     - if selected intervals intersect
+       - and their right borders are equal, put their intersection to the accumulator
+         and proceed to the next intervals in both lists
+       - and their right borders are not equal, put their intersection to the accumulator and truncate the interval
+         with the least right border*)
+  let rec meet_lists ?(acc = []) l1 l2 =
```

Comment:
I think this relies that within one set there is not overlap between contained intervals, are you actually enforcing this?

Reply:
Yes, because overlap might happen only as a result of an arithmetic operation, and for these I have `merge_intersect`.